plugins/by-name/typescript-tools/default.nix

```diff
@@ -157,38 +157,43 @@ in
       };
     };
   };
-  config = mkIf cfg.enable {
-    extraPlugins = [ cfg.package ];
 
-    plugins.lsp.postConfig =
-      with cfg;
+  # NOTE: call setup in lsp config
+  callSetup = false;
+  extraConfig = cfg: {
+    plugins.lsp.postConfig = ''
+      require('typescript-tools').setup(${lib.nixvim.toLuaObject cfg.settings})
+    '';
+
+    warnings =
       let
-        options = {
-          inherit handlers;
-          on_attach = onAttach;
-
-          settings = with settings; {
-            separate_diagnostic_server = separateDiagnosticServer;
-            publish_diagnostic_on = publishDiagnosticOn;
-            expose_as_code_action = exposeAsCodeAction;
-            tsserver_path = tsserverPath;
-            tsserver_plugins = tsserverPlugins;
-            tsserver_max_memory = tsserverMaxMemory;
-            tsserver_format_options = tsserverFormatOptions;
-            tsserver_file_preferences = tsserverFilePreferences;
-            tsserver_locale = tsserverLocale;
-            complete_function_calls = completeFunctionCalls;
-            include_completions_with_insert_text = includeCompletionsWithInsertText;
-            code_lens = codeLens;
-            disable_member_code_lens = disableMemberCodeLens;
-            jsx_close_tag = with jsxCloseTag; {
-              inherit enable filetypes;
-            };
-          };
-        };
+        definedOpts = lib.filter (opt: lib.hasAttrByPath (lib.toList opt) cfg.settings) [
+          "separateDiagnosticServer"
+          "publishDiagnosticOn"
+          "exposeAsCodeAction"
+          "tsserverPath"
+          "tsserverPlugins"
+          "tsserverMaxMemory"
+          "tsserverFormatOptions"
+          "tsserverFilePreferences"
+          "tsserverLocale"
+          "completeFunctionCalls"
+          "includeCompletionsWithInsertText"
+          "codeLens"
+          "disableMemberCodeLens"
+          [
+            "jsxCloseTag"
+            "enable"
+          ]
+          [
+            "jsxCloseTag"
+            "filetypes"
+          ]
+        ];
       in
-      ''
-        require('typescript-tools').setup(${helpers.toLuaObject options})
+      lib.optional (definedOpts != [ ]) ''
+        Nixvim(plugins.typescript-tools): The following settings have moved under `plugins.typescript-tools.settings.settings` with snake_case:
+        ${lib.concatMapStringsSep "\n" (opt: "  - ${lib.showOption (lib.toList opt)}") definedOpts}
       '';
```

Comment:
In addition to this warning, it'd be nice to _try_ and have some migration aliases...

Defining `settings.settings` conditionally on whether or not one of the old attrs is present would be infinitely recursive, _however_ I think we can work around that by always defining `settings.settings`, and having the _content_ be dependent on whether the old attrs are in `cfg.settings`.

Something like:
```nix
# Psudonix
settings.settings = lib.optionalAttrs (definedOpts != [ ]) (
  fold (
    acc: oldName:
    recursiveUpdate acc (
      setAttrByPath (toSnakeCasePath oldName) (getAttrFromPath oldName cfg.settings)
    )
  ) { } definedOpts
)
```

Also: we should check that the plugin itself doesn't care that we're passing additional unknown table entries through to its `setup` function.

Reply:
That's what i'm running into, trying to see if I can get around infinite recursion.

Reply:
Thanks for trying! If it doesn't work, don't sweat too hard over it. I knew `attrsOf` had some inf-recursion limitations here, I just wasn't sure how it'd be affected if the conditional stuff was "a level above" the attr being defined...

My curiosity is itching to figure out a solution, but it isn't worth laboring over when the warnings alone are _good enough_.

Reply:
Yeah... I really wanted to get it to work because it'd be one less thing for people to have to do. But, it's feeling a bit above my skillset, here.

Reply:
I just realised, because we can't rename these to the correct names, we are actually _breaking_ users configs.

I also had a thought about how we can make the renames _kinda_ work:

Maybe instead of doing the recursive `settings.settings = mapToNewSettings cfg.settings`, we could fallback to our "old school" _set `settingsOptions` in a let-in block before calling `toLuaObject`_ technique.

This would bypass the infinite recursion, because we wouldn't be assigning to the `settings` option.